File: 3-1.ds stack/src/cppsrc/mystack.cpp

#ifndef __MYSTACK_CPP_
#define __MYSTACK_CPP_

#include <iostream>

template <typename T>
class MyStack {
private:
    T* data = nullptr;
    size_t cursor = 0;
    const size_t capacity = 0;

public:
    MyStack(size_t N) : capacity(N)
    {
        data = new T[N];
    }

    ~MyStack()
    {
        delete data;
    }

    bool isEmpty()
    {
        return cursor == 0;
    }

    bool isFull()
    {
        return cursor == capacity;
    }

    void push(T val)
    {
        if (isFull()) {
            throw "stack is full";
        }
        data[cursor++] = val;
    }

    void pop()
    {
        if (isEmpty()) {
            throw "stack is empty";
        }
        --cursor;
    }

    T top()
    {
        if (isEmpty()) {
            throw "stack is empty";
        }
        return data[cursor - 1];
    }
};

#endif

File: 3-1.ds stack/src/mystack.ts
type TypedArray = Int8Array | Int16Array | Int32Array | Uint8Array | Uint16Array | Uint32Array;
type TypedArrayConstructor = {
    new (size: number): TypedArray;
};

export class MyStack {
    private typedArray: TypedArray;
    private size: number = 0;
    private capacity: number = 0;
    constructor(typedArray: TypedArrayConstructor, size: number) {
        this.capacity = size;
        this.typedArray = new typedArray(this.capacity);
    }

    isEmpty() {
        return this.size === 0;
    }

    isFull() {
        return this.size === this.capacity;
    }

    push(val: number) {
        if (this.isFull()) {
            throw new Error("stack is full");
        }
        this.typedArray[this.size++] = val;
    }

    pop() {
        if (this.isEmpty()) {
            throw new Error("deque is empty");
        }
        --this.size;
    }

    top(): number {
        if (this.isEmpty()) {
            throw new Error("deque is empty");
        }
        return this.typedArray[this.size - 1];
    }
}
